tests-e2e/receipts_test.go

```diff
@@ -222,3 +222,116 @@ func TestReceiptsPrivate(t *testing.T) {
+	room4 := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat", "name": "room 2"})
+	t.Logf("room1=%s room2=%s room3=%s room4=%s", room1, room2, room3, room4)
+
+	t.Log("Bob joins those rooms.")
+	bob.JoinRoom(t, room1, nil)
+	bob.JoinRoom(t, room2, nil)
+	bob.JoinRoom(t, room3, nil)
+	bob.JoinRoom(t, room4, nil)
+
+	t.Log("Alice posts a message to each rooms")
+	messageEvent := Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello, room!",
+		},
+	}
+	message1 := alice.SendEventSynced(t, room1, messageEvent)
+	message2 := alice.SendEventSynced(t, room1, messageEvent)
+	message3 := alice.SendEventSynced(t, room1, messageEvent)
+	message4 := alice.SendEventSynced(t, room1, messageEvent)
+
+	t.Log("Bob posts a public read receipt for the messages in rooms 1 and 2.")
+	bob.SendReceipt(t, room1, message1, "m.read")
+	bob.SendReceipt(t, room2, message2, "m.read")
+	time.Sleep(300 * time.Millisecond) // TODO: find a better way to wait until the proxy has processed this.
+
+	t.Log("Bob makes an initial sliding sync, requesting receipts in rooms 2 and 4only.")
+	syncResp = bob.SlidingSync(t, sync3.Request{
+		Extensions: extensions.Request{
+			Receipts: &extensions.ReceiptsRequest{
+				Core: extensions.Core{Enabled: &boolTrue, Lists: []string{}, Rooms: []string{room2, room4}},
+			},
+		},
+		Lists: map[string]sync3.RequestList{
+			"window": {
+				Ranges: sync3.SliceRanges{{0, 20}},
+			},
+		},
+	})
+
+	t.Log("Bob should see his receipt in room 2, but not his receipt in room 1.")
+	m.MatchResponse(
+		t,
+		syncResp,
+		m.MatchReceipts(room1, nil),
+		m.MatchReceipts(room2, []m.Receipt{{
+			EventID: message2,
+			UserID:  bob.UserID,
+			Type:    "m.read",
+		}}),
+	)
+
+	responses := make(chan *sync3.Response, 10)
+	pos := syncResp.Pos
+	bobIncrementalSync := func() {
+		t.Log("Bob requests an incremental sync...")
+		response := bob.SlidingSync(
+			t,
+			sync3.Request{
+				Extensions: extensions.Request{
+					AccountData: &extensions.AccountDataRequest{
+						Core: extensions.Core{Enabled: &boolTrue, Lists: []string{}, Rooms: []string{room2}},
+					},
+				},
+			},
+			WithPos(pos),
+		)
+		pos = response.Pos
+		responses <- response
+	}
+	waitForSyncResponse := func() *sync3.Response {
+		t.Log("... Bob waits for the incremental sync response...")
+		select {
+		case res := <-responses:
+			return res
+		case <-time.After(500 * time.Millisecond):
+			t.Fatalf("Timed out waiting for incremental sync response")
+		}
+		return nil
+	}
+
+	go bobIncrementalSync()
+	t.Log("Bob posts receipts for rooms 3 and 4.")
+	bob.SendReceipt(t, room3, message3, "m.read")
+	bob.SendReceipt(t, room4, message4, "m.read")
+	time.Sleep(300 * time.Millisecond) // TODO: find a better way to wait until the proxy has processed this.
+
+	syncResp = waitForSyncResponse()
+	t.Log("Bob sees his receipt in room 4, but not in room 3.")
```

Comment:
So testing for negation I would do what you're doing here and send room 3 BEFORE room 4. That way, you just need to test until you see room 4 and not wait an indeterminate amount of time for room 3, effectively using room 4 as a sentinel.